src/3d/chunks/qgschunkloader_p.cpp

```diff
@@ -15,10 +15,50 @@
 
 #include "qgschunkloader_p.h"
 #include "qgschunknode_p.h"
+#include "qgscoordinatetransform.h"
+#include "qgs3dutils.h"
+#include "qgsvectorlayer.h"
 
 #include <QVector>
 ///@cond PRIVATE
 
+
+// ===============================
+// QgsChunkLoader
+// ===============================
+
+void QgsChunkLoader::buildVectorFeatureRequest( const QgsVectorLayer *layer, const QgsChunkNode *node, const Qgs3DMapSettings &mapSettings, const QSet<QString> &attributeNames, QgsFeatureRequest &request )
+{
+  request.setDestinationCrs( mapSettings.crs(), mapSettings.transformContext() );
+  request.setSubsetOfAttributes( attributeNames, layer->fields() );
+
+  // only a subset of data to be queried
+  // data which are inside the node and the extent of the scene
+  QgsRectangle bboxExtent = Qgs3DUtils::worldToMapExtent( node->bbox(), mapSettings.origin() );
+  QgsGeometry sceneExtentGeom = mapSettings.orientedExtent();
+  if ( mapSettings.crs() != layer->crs() )
+  {
+    QgsCoordinateTransform ct( mapSettings.crs(), layer->crs(), mapSettings.transformContext() );
+    ct.setBallparkTransformsAreAppropriate( true );
+    try
+    {
+      bboxExtent = ct.transformBoundingBox( bboxExtent );
+      sceneExtentGeom.transform( ct );
+    }
+    catch ( QgsCsException & )
+    {
+      QgsDebugError( QStringLiteral( "Error transforming node bounds coordinate" ) );
+    }
+  }
+  QString filterExpression = QStringLiteral( "intersects($geometry, geomFromWkt('%1')) AND intersects($geometry, geomFromWkt('%2'))" ).arg( sceneExtentGeom.asWkt(), bboxExtent.asWktPolygon() );
```

Comment:
This approach is a performance killer. It will force a full fetch of the entire layer's features, and then an evaluation on the QGIS client side for the intersection. I.e. it will negate any potential use of the backend's spatial index.

There's two options:

1. Add a geometry filter option to QgsFeatureRequest, and implement for all provider iterators. (Warning: This is a LOT of work.)
2. Use the bounding box of the geometry as a filterRect for the request, and then do the intersection test manually for each fetched feature, using a prepared GEOS geometry of the desired polygon object.

Reply:
Thanks. I have done the second option.